garage/tf/algos/npo.py

```diff
@@ -153,3 +137,274 @@ def get_itr_snapshot(self, itr, samples_data):
+                policy_state_info_vars_flat = flatten_batch_dict(
+                    policy_state_info_vars, name="policy_state_info_vars_flat")
+                policy_old_dist_info_vars_flat = flatten_batch_dict(
+                    policy_old_dist_info_vars,
+                    name="policy_old_dist_info_vars_flat")
+
+            # valid view
+            with tf.name_scope("valid"):
+                action_valid = filter_valids(
+                    action_flat, valid_flat, name="action_valid")
+                policy_state_info_vars_valid = filter_valids_dict(
+                    policy_state_info_vars_flat,
+                    valid_flat,
+                    name="policy_state_info_vars_valid")
+                policy_old_dist_info_vars_valid = filter_valids_dict(
+                    policy_old_dist_info_vars_flat,
+                    valid_flat,
+                    name="policy_old_dist_info_vars_valid")
+
+        # policy loss and optimizer inputs
+        pol_flat = namedtuple_singleton(
+            "PolicyLossInputsFlat",
+            obs_var=obs_flat,
+            action_var=action_flat,
+            reward_var=reward_flat,
+            valid_var=valid_flat,
+            policy_state_info_vars=policy_state_info_vars_flat,
+            policy_old_dist_info_vars=policy_old_dist_info_vars_flat,
+        )
+        pol_valid = namedtuple_singleton(
+            "PolicyLossInputsValid",
+            action_var=action_valid,
+            policy_state_info_vars=policy_state_info_vars_valid,
+            policy_old_dist_info_vars=policy_old_dist_info_vars_valid,
+        )
+        policy_loss_inputs = namedtuple_singleton(
+            "PolicyLossInputs",
+            obs_var=obs_var,
+            action_var=action_var,
+            reward_var=reward_var,
+            baseline_var=baseline_var,
+            valid_var=valid_var,
+            policy_state_info_vars=policy_state_info_vars,
+            policy_old_dist_info_vars=policy_old_dist_info_vars,
+            flat=pol_flat,
+            valid=pol_valid,
+        )
+        policy_opt_inputs = namedtuple_singleton(
+            "PolicyOptInputs",
+            obs_var=obs_var,
+            action_var=action_var,
+            reward_var=reward_var,
+            baseline_var=baseline_var,
+            valid_var=valid_var,
+            policy_state_info_vars_list=policy_state_info_vars_list,
+            policy_old_dist_info_vars_list=policy_old_dist_info_vars_list,
+        )
+
+        return policy_loss_inputs, policy_opt_inputs
+
+    def _build_policy_loss(self, i):
+        pol_dist = self.policy._dist
+
+        policy_entropy = self._build_entropy_terms(i)
+
+        with tf.name_scope("augmented_rewards"):
+            rewards = i.reward_var \
+                      + self.policy_ent_coeff * policy_entropy
+
+        with tf.name_scope("policy_loss"):
+            with tf.name_scope("advantage"):
+                gamma_lambda = tf.constant(
```

Comment:
can you please move the advantage calculation into a function tensor_utils? it's a very subtle formulation which is easy to get wrong (it took me about a day to write). Please also retain my block comment explaining the method.

it should probably also be tested.

Reply:
em, how should i test it?

Reply:
a simple unit test with known correct input and output would suffice.